Sources/SwiftSyntax/SyntaxNodes.swift.gyb

```diff
@@ -159,6 +159,57 @@ public struct ${node.name}: ${base_type}, _SyntaxBase, Hashable {
   }
 %     end
 
+  /// Returns a new `${node.name}` with its leading trivia replaced
+  /// by the provided trivia.
+  public func withLeadingTrivia(_ leadingTrivia: Trivia) -> ${node.name} {
+    return ${node.name}(data.withLeadingTrivia(leadingTrivia))
+  }
+
+  /// Returns a new `${node.name}` with its trailing trivia replaced
+  /// by the provided trivia.
+  public func withTrailingTrivia(_ trailingTrivia: Trivia) -> ${node.name} {
+    return ${node.name}(data.withTrailingTrivia(trailingTrivia))
+  }
+
+  /// Returns a new `${node.name}` with its leading trivia removed.
+  public func withoutLeadingTrivia() -> ${node.name} {
+    return withLeadingTrivia([])
+  }
+
+  /// Returns a new `${node.name}` with its trailing trivia removed.
+  public func withoutTrailingTrivia() -> ${node.name} {
+    return withTrailingTrivia([])
+  }
+
+  /// Returns a new `${node.name}` with all trivia removed.
+  public func withoutTrivia() -> ${node.name} {
+    return withoutLeadingTrivia().withoutTrailingTrivia()
+  }
+
+  /// The leading trivia (spaces, newlines, etc.) associated with this `${node.name}`.
+  public var leadingTrivia: Trivia? {
+    get {
+      return raw.formLeadingTrivia()
+    }
+    set {
+      if let leadingTrivia = newValue {
+        self = withLeadingTrivia(leadingTrivia)
+      }
+    }
+  }
+
+  /// The trailing trivia (spaces, newlines, etc.) associated with this `${node.name}`.
+  public var trailingTrivia: Trivia? {
+    get {
+      return raw.formTrailingTrivia()
+    }
+    set {
+      if let trailingTrivia = newValue {
+        self = withTrailingTrivia(trailingTrivia)
```

Comment:
What if I set this to `nil`?

Reply:
This is a tricky one. As the getter returns `nil` when there is no token in the hierarchy, it doesn't make sense to set nil: it's not like it should remove all tokens form the hierarchy. The only options I thought of are:

- Have the setter do nothing when setting nil (that's the option I chose). Unfortunately, this can be very surprising:

```swift
print(node.leadingTrivia) // [.spaces(1)]
node.leadingTrivia = nil
print(node.leadingTrivia) // [.spaces(1)]
```

- Have the setter trap when setting nil. That's another good option AFAIK.
- Don't implement a setter and only rely on `with...Trivia` functions. I'm not a fan of this option as we have convenience setter for all other properties.

Reply:
What about making `node.leadingTrivia = nil` have the same semantics as setting `node.leadingTrivia = []`? Without looking too much into detail, it's a behaviour I would expect / would be able to explain when it occurs.

Reply:
That's indeed another option, but it still has the same surprising behaviour as noop that the getter won't be returning the same value you just set.

Reply:
Yeah, still both values represent some kind of “nothingness”. 

It also reminded me of some Cocoa pattern where setting a property to `nil` caused it to take on a particular value (I think like black for text colour). It's not really relevant here but it shows that having the setter set a different value than is afterwards returned by the getter is not completely unprecedented in the Apple ecosystem.

Reply:
Sounds fair, I'll make the change.

Reply:
FWIW that pattern is codified into the language as `null_resettable`. I think it's perfectly valid here.